### PR TITLE
Refactor server orders

### PR DESCRIPTION
**Pull request by teinarss**
Changing on how the disconnect order is sent from the server, ie not relying on spoofing the from client. Also did some cleanup around dispatching server orders.

### CODE REVIEW

File: OpenRA.Game/Server/Server.cs
Comment (pchote) [inline OpenRA.Game/Server/Server.cs]: Can we please use "Disconnect" here and above for consistency?

File: OpenRA.Game/Server/Server.cs
Comment (pchote) [inline OpenRA.Game/Server/Server.cs]: Stray `@`?

File: OpenRA.Game/Network/Order.cs
Comment (jrb0001) [inline OpenRA.Game/Network/Order.cs]: This will need an increase in the protocol version.

File: OpenRA.Game/Network/Order.cs
Comment (jrb0001) [inline OpenRA.Game/Network/Order.cs]: What is this ExtraData used for?

Comment (teinarss) [reply OpenRA.Game/Network/Order.cs]: Used for sending the ClientId.

Comment (jrb0001) [reply OpenRA.Game/Network/Order.cs]: S->C is always Type+Length+ClientId+FramePayload so you already have the ClientId there. You don't want it in the C->S direction anyway, especially if the goal of this PR is to avoid spoofing.

Comment (teinarss) [reply OpenRA.Game/Network/Order.cs]: Not in this case, this is sent from the server, before the server did spoof this as sent from the Client. See https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Game/Server/Server.cs#L1042

Comment (jrb0001) [reply OpenRA.Game/Network/Order.cs]: Then I don't really understand the goal of this PR. I don't really see the current logic as spoofing, because the server simply translate the EOF (which comes from the client, directly or indirectly) into something that can be sent to other clients. It is also a dedicated frame type (and not an order) and those are meant for such special cases in my opinion.

Note that there is (was?) also a Disconnect order which was sent as a normal order over the connection. I think it would be much better to remove that one instead, so all disconnect situations work exactly the same way and avoid the *order* spoofing situation.

Comment (teinarss) [reply OpenRA.Game/Network/Order.cs]: Calling it spoofing was maybe using the wrong word. more like stop faking that the disconnect order is sent from the client, and the goal of this PR is to clean up the network code.

Comment (jrb0001) [reply OpenRA.Game/Network/Order.cs]: How do the other clients know when (which frame) the disconnect happened? Server orders are frame=0 by definition and you don't have a dedicated field for it.

Comment (jrb0001) [reply OpenRA.Game/Network/Order.cs]: I have always seen the protocol like this, and then the current design makes sense (except for the disconnect *order* which was introduced later):
- A client can only send stuff from itself, so the ClientId field is omitted for C->S. Otherwise it works the same as S->C.
- The server basically bundles many logical streams (one per client plus one for the server) into a single connection:
  - The server is just another client, but with ClientId=0.
  - Frames with Frame=0 are to be processed immediately.
  - All other frames form an ordered stream based on Frame!=0, which must not have any gaps or duplicates.
  - A new stream is implicitely opened if the ClientId hasn't been seen before.
  - A stream is closed explicitely with a Disconnect frame (plus a Disconnect order which is weird).
  - In case of replays the Disconnect frame and order are ignored and it is derived from the last encountered order (weird).
  - SyncHash is optional data and ordered independently. It would probably make more sense to set Frame=0 and include the real frame number inside the payload? On the other hand it is also a stream, just independent from the order stream.

Comment (teinarss) [reply OpenRA.Game/Network/Order.cs]: >  How do the other clients know when (which frame) the disconnect happened? Server orders are frame=0 by definition and you don't have a dedicated field for it.
>

It was never needed, and the code was just weird and was removed in a previous PR.

> 
> I have always seen the protocol like this, and then the current design makes sense (except for the disconnect _order_ which was introduced later):
> 
>     * A client can only send stuff from itself, so the ClientId field is omitted for C->S. Otherwise it works the same as S->C.
> 
>     * The server basically bundles many logical streams (one per client plus one for the server) into a single connection:
>       
>       * The server is just another client, but with ClientId=0.

Correct.

>       * Frames with Frame=0 are to be processed immediately.

Correct.

>       * All other frames form an ordered stream based on Frame!=0, which must not have any gaps or duplicates.
>       * A new stream is implicitely opened if the ClientId hasn't been seen before.

Not anymore, now these are created when the games is started.

>       * A stream is closed explicitely with a Disconnect frame (plus a Disconnect order which is weird).

In the future i want to remove one of these orders.

>       * In case of replays the Disconnect frame and order are ignored and it is derived from the last encountered order (weird).

Something for a future cleanup.

>       * SyncHash is optional data and ordered independently. It would probably make more sense to set Frame=0 and include the real frame number inside the payload? On the other hand it is also a stream, just independent from the order stream.

**Comment by teinarss** [comment]
Updated, not sure how we want to handle protocol version, increase it in every PR that changes stuff or in one final PR after all these PR have been merged?

File: OpenRA.Game/Server/Server.cs
Comment (pchote) [inline OpenRA.Game/Server/Server.cs:770]: The current code only ever sends one order, so how about
```csharp
public void DispatchServerOrderToClients(Order order)
{
	var data = order.Serialize();
```
to avoid all the callers having to Serialize?

**Comment by jrb0001** [comment]
If you don't increase the protocol version in every PR then thirdparty mods might run into a version/implementation mismatch which can lead to strange problems. So it would be best to either merge all PRs at the same time or to increase it in every PR.

**Comment by abcdefg30** [comment]
Tbh I still think tools depending on bleed versions is kind of an anti-feature, but our order version is an int which means we can be generous with increasing it.

**Comment by teinarss** [comment]
Updated

File: OpenRA.Game/Network/Order.cs
Comment (pchote) [inline OpenRA.Game/Network/Order.cs]: `var clientID = `

File: OpenRA.Game/Network/Order.cs
Comment (pchote) [inline OpenRA.Game/Network/Order.cs]: Check that this value is "Disconnect" and throw if it isn't?

File: OpenRA.Game/Network/OrderManager.cs
Comment (pchote) [inline OpenRA.Game/Network/OrderManager.cs]: Can we define a const `Order.DisconnectOrderLength = 16` like we do for `SyncHashOrderLength` and then check it here as `packet.Length == Order.DisconnectOrderLength + 4`?

File: OpenRA.Game/Network/OrderManager.cs
Comment (pchote) [inline OpenRA.Game/Network/OrderManager.cs]: `var clientID = (int)packet.ToOrderList(World).First().ExtraData;`

File: OpenRA.Game/Server/Server.cs
Comment (pchote) [inline OpenRA.Game/Server/Server.cs]: Rider is saying that the verbatim prefixes here and below are redundant, so can we please drop these?

**Review by pchote — COMMENTED** [review]
The change from spoofing the disconnect order sender to having it explicitly sent by the server seems like a good idea. I don't see the advantage for wrapping it in a serialized Order though, this seems to add a lot of complexity and redundancy compared to making `OrderType.Disconnect` carry just the clientID the same way `OrderType.SyncHash` carries the sync state.

If we do want to keep the `Order` wrapping can we please change some of my suggestions below to make the code a bit more self explanatory?

**Comment by teinarss** [comment]
Updated.

File: OpenRA.Game/Server/Server.cs
Comment (pchote) [inline OpenRA.Game/Server/Server.cs]: The `clientId` is an int that keeps incrementing as players join and leave the server. Its entirely possible that we might overflow this byte, and `DisconnectOrderLength` also assumes that we are using an int for the packet to be handled at all!

How about
```csharp
var disconnectPacket = new MemoryStream(5);
disconnectPacket.WriteByte((byte)OrderType.Disconnect);
disconnectPacket.Write(toDrop.PlayerIndex);
DispatchServerOrdersToClients(disconnectPacket.ToArray());
```
...

File: OpenRA.Game/Network/OrderManager.cs
Comment (pchote) [inline OpenRA.Game/Network/OrderManager.cs]: ...and then `BitConverter.ToInt32(packet, 5)` here.

Also, on closer inspection my suggestion for calling this `clientID` wasn't great because we already have `clientId` above representing the sending client (the server). So maybe call this `disconnectedClient`? Or now that we aren't unwrapping the Order list its probably clear enough to go back to to `pendingPackets.Remove(BitConverter.ToInt32(packet, 5))`?

File: OpenRA.Game/Server/Server.cs
Comment (pchote) [inline OpenRA.Game/Server/Server.cs:765]: Missed the "CreateGameSave" case on line 908 when converting the `DispatchOrders(null, 0, ...`s to this.

File: OpenRA.Game/Server/Server.cs
Comment (pchote) [inline OpenRA.Game/Server/Server.cs]: Missed this `@`

File: OpenRA.Game/Server/Server.cs
Comment (pchote) [inline OpenRA.Game/Server/Server.cs]: And this one. Maybe drop the linebreak too?

File: OpenRA.Game/Network/OrderManager.cs
Comment (pchote) [inline OpenRA.Game/Network/OrderManager.cs:134]: It's weird that the server code treats `OrderType` as the first byte in the data payload rather than as part of the framing. Fixing this isn't trivial because the `OrderType` handling is split half between the network code and half inside `Order`, so best that we leave this to a future PR.

**Comment by teinarss** [comment]
Updated

File: OpenRA.Game/Server/Server.cs
Comment (jrb0001) [reply OpenRA.Game/Server/Server.cs]: Agreed, an older version of my server implementation allocated numbers with at least 24 significant bits by default and it worked perfectly. Limiting the disconnect to just a signed byte just makes no sense as long as it uses the full int everywhere else.

**Review by jrb0001 — DISMISSED** [review]
The protocol implementation looks fine to me.

File: OpenRA.Game/Server/Server.cs
Comment (pchote) [inline OpenRA.Game/Server/Server.cs:740]: Good thing I tested before merging! This looked safe to me, but it turns out that we do still have at least one case where `conn` is null - `SendMessage` below defaults to it in most cases. Not sure if you want to revert this line or change `SendMessage` to use `DispatchServerOrdersToClients`?

Comment (pchote) [reply OpenRA.Game/Server/Server.cs:740]: LGTM and testing revealed no further obvious problems after reverting this line.

**Comment by teinarss** [comment]
Updated

**Review by pchote — DISMISSED** [review]

File: OpenRA.Game/Server/Server.cs
Comment (pchote) [inline OpenRA.Game/Server/Server.cs]: I think the `conn != null` check here and just below can be removed now too?

**Comment by teinarss** [comment]
Updated.

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/7b5063a8192f077a86281d9364ffc497f956e81a

**Pull request merged**
